cmd/kubeadm/app/phases/upgrade/compute.go

```diff
@@ -125,11 +151,14 @@ func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesA
 
 	// Construct a descriptor for the current state of the world
 	beforeState := ClusterState{
-		KubeVersion:     clusterVersionStr,
-		DNSVersion:      dnsVersion,
-		KubeadmVersion:  kubeadmVersionStr,
-		KubeletVersions: kubeletVersions,
-		EtcdVersion:     etcdVersion,
+		KubeVersion:                   clusterVersionStr,
```

Comment:
If run `kubeadm upgrade plan` on different master nodes or with a `--kubeconfig` flag which access different master nodes via LB, the output may be different. Should we consider this as a bug? i.e. a cluster with 3 master nodes, one is v1.7, one is v1.8, and one is v1.9.

Can we choose the master node with the lowest version as cluster version?

/cc @neolit123 @pacoxu

Reply:
ClusterVersion comes from the discovery client ServerVersion of a server it finds at the LB.
https://github.com/kubernetes/kubernetes/pull/123578/files#diff-6677166aabde765ebd3c6cc028c39e0b22fe33defc8fc45c2ebb972608f3ae9dR93
https://github.com/carlory/kubernetes/blob/46c9bd12674fa502ca33813e06a4323a1f485581/cmd/kubeadm/app/phases/upgrade/versiongetter.go#L76

it's true that an admin.conf can reach a different version API servers and the version can be different.

a potential change in kubeadm can be the following:
- ignore the discovery ServerVersion
- get all API server versions in the cluster
- use the latest one as the "cluster version"

WDYT?

Reply:
if we use the latest, we may miss some available upgrade plans. how about the oldest one?

Reply:
from the Kubernetes perspective the latest apiserver version makes sense because that is where all the new APIs and removed old APIs will happen.
however i can see that from the kubeadm upgrade perspective the oldes version makes more sense so that we can list all the upgrades.

as you can understand in a cluster "KubeVersion" doesn't make sense because of different apiservers and kubelets and ....
so if we want to catch all the upgrades we should probably rename KubeVersion -> OldestKubeVersion and make OldestKubeVersion be the oldest apiserver version.

Reply:
If apiserver versions are difference(minor version: 1.9 vs 1.8), I think we should return error and a message that suggests user upgrades their low version apiserver to newer. 

If  apiserver versions are difference in patch versions, I think we can take the latest version of all apiservers.

Reply:
> If apiserver versions are difference(minor version: 1.9 vs 1.8), I think we should return error and a message that suggests user upgrades their low version apiserver to newer.

i think it can be just a warning:
```
klog.Warningf("Different API server versions in the cluster were discovered: %v. Please upgrade your control plane" +
" nodes to the same version of Kubernetes", someVersionSlice)
```

if we decided this is a problem for more users we can make it an error.

i have seen users with different apiserver vertsions by mistakes, but not many.

Reply:
Edited my comments above. 

A warning is needed.